### DISCUSSION TITLE
Examples of using UT-Assert with c++ cFS application

### DISCUSSION THREAD

**Discussion by aliounis** [nasa/cFS#527] [Q&A]
Hi all,

Does anyone have open examples of cFS apps written using primarily c++ that also use ut-assert for unit testing?  I'm just starting to investigate this for an app being developed and want to see if there are any gotchas/issues with using the combination.

**Comment by aliounis** [comment]
For anyone else interested in this, I found no examples, but have come across one main issue/solution.

The issue is that the normal stubbing functionality for cFS does not play nicely with class methods in my experience.  The way to work around this is to create stub free functions, call these free functions from the stubbed method, and then create macros to make it easy to get the free function that corresponds to each method that has been stubbed. I've created a script to do this automatically (similar to the generate_stubs.pl) and hope to share it at some point (once I have time to usher it through the NASA software release process).

**Reply by skliper** [reply]
The UT_KEY part really is just trying to get a unique identifier for the related UT elements to do look-ups and assignments.  You can replace using the function pointers with any alternative "key" method as long as it's unique across the test.

EDIT: not sure if this is related, but another C++ user mentioned this challenge

**Reply by aliounis** [reply]
Ok, yeah I see, so essentially there just needs to be a name (or pointer) -> unique integer matching scheme that could override the default use of the integer cast of the function pointer.

Regardless, I needed a new script to auto generate the stubs for c++, since the perl script didn't play nicely with c++ methods, so the work around I came up with wasn't too big of a hassle, but it would probably be cleaner to do it this way.

**Reply by ivanperez-keera** [reply]
@aliounis Were you ever able to release it? If not, can you send it to me via NASA email? You should be able to find me in the directory as Ivan Perez Dominguez. Thanks!
